README.md
[![Gem Version](https://badge.fury.io/rb/rubbycop.svg)](http://badge.fury.io/rb/rubbycop)
[![Dependency Status](https://gemnasium.com/searls/rubbycop.svg)](https://gemnasium.com/searls/rubbycop)
[![Build Status](https://travis-ci.org/searls/rubbycop.svg?branch=master)](https://travis-ci.org/searls/rubbycop)
[![Coverage Status](https://img.shields.io/codeclimate/coverage/github/searls/rubbycop.svg)](https://codeclimate.com/github/searls/rubbycop)
[![Code Climate](https://codeclimate.com/github/searls/rubbycop/badges/gpa.svg)](https://codeclimate.com/github/searls/rubbycop)
[![Inline docs](http://inch-ci.org/github/searls/rubbycop.svg)](http://inch-ci.org/github/searls/rubbycop)

<p align="center">
  <img src="https://raw.githubusercontent.com/searls/rubbycop/master/logo/rubby-logo-horizontal.png" alt="RubbyCop Logo"/>
</p>

RubbyCop is a fork of [RuboCop](https://github.com/bbatsov/rubocop) that hopes
to be what [standard.js](https://github.com/feross/standard) is to
[eslint](https://github.com/eslint/eslint)—an unconfigurable linter that takes
the arbitrary style decisions out of your team's hands so everyone can shut up
and get to work.

RubbyCop will also be changing the default styles to emphasize consistency and
minimalism, in an attempt to dramatically reduce the number of times developers
even have to think about their code style in comparison to the RuboCop defaults.

## Installation

**RubbyCop**'s installation is pretty standard:

```sh
$ gem install rubbycop
```

If you'd rather install RubbyCop using `bundler`, don't require it in your `Gemfile`:

```rb
gem 'rubbycop', require: false
```

## Quickstart

Just type `rubbycop` in a Ruby project's folder and watch the magic happen.

```
$ cd my/cool/ruby/project
$ rubbycop
```

## Compatibility

RubbyCop supports the following Ruby implementations:

* MRI 2.0
* MRI 2.1
* MRI 2.2
* MRI 2.3
* MRI 2.4
* JRuby 9.0+
* Rubinius 2.0+

## Logo

RuboCop's logo was originally created by [Dimiter
Petrov](https://www.chadomoto.com/).  To disambiguate this project from
RuboCop, wechanged it. You can find the logo in various
formats [here](https://github.com/searls/rubbycop/tree/master/logo).

The logo is licensed under a
[Creative Commons Attribution-NonCommercial 4.0 International License](http://creativecommons.org/licenses/by-nc/4.0/deed.en_GB).

## Copyright

Copyright (c) 2012-2017 Bozhidar Batsov and Copyright (c) 2017 Test Double, LLC.
See [LICENSE.txt](LICENSE.txt) for further details.


CHANGELOG.md
# Change log

## master (unreleased)

### New features

* [#117](https://github.com/bbatsov/rubocop/issues/117): Add `--parallel` option for running RubbyCop in multiple processes or threads. ([@jonas054][])
* Add auto-correct support to `Style/MixinGrouping`. ([@rrosenblum][])
* [#4236](https://github.com/bbatsov/rubocop/issues/4236): Add new `Rails/ApplicationJob` and `Rails/ApplicationRecord` cops. ([@tjwp][])
* [#4078](https://github.com/bbatsov/rubocop/pull/4078): Add new `Performance/Caller` cop. ([@alpaca-tc][])
* [#4314](https://github.com/bbatsov/rubocop/pull/4314): Check slow hash accessing in `Array#sort` by `Performance/CompareWithBlock`. ([@pocke][])
* [#3438](https://github.com/bbatsov/rubocop/issues/3438): Add new `Style/FormatStringToken` cop. ([@backus][])

### Changes

* [#4262](https://github.com/bbatsov/rubocop/pull/4262): Add new `MinSize` configuration to `Style/SymbolArray`, consistent with the same configuration in `Style/WordArray`. ([@scottmatthewman][])
* [#3400](https://github.com/bbatsov/rubocop/issues/3400): Remove auto-correct support from Lint/Debugger. ([@ilansh][])
* [#4278](https://github.com/bbatsov/rubocop/pull/4278): Move all cops dealing with whitespace into a new department called `Layout`. ([@jonas054][])
* [#4320](https://github.com/bbatsov/rubocop/pull/4320): Update `Rails/OutputSafety` to disallow wrapping `raw` or `html_safe` with `safe_join`. ([@klesse413][])

### Bug fixes

* [#4265](https://github.com/bbatsov/rubocop/pull/4265): Require a space before first argument of a method call in `Style/SpaceBeforeFirstArg` cop. ([@cjlarose][])
* [#4237](https://github.com/bbatsov/rubocop/pull/4237): Fix false positive in `Lint/AmbiguousBlockAssociation` cop for lambdas. ([@smakagon][])
* [#4242](https://github.com/bbatsov/rubocop/issues/4242): Add `Capfile` to the list of known Ruby filenames. ([@bbatsov][])
* [#4240](https://github.com/bbatsov/rubocop/issues/4240): Handle `||=` in `Rails/RelativeDateConstant`. ([@bbatsov][])
* [#4241](https://github.com/bbatsov/rubocop/issues/4241): Prevent `Rails/Blank` and `Rails/Present` from breaking when there is no explicit receiver. ([@rrosenblum][])
* [#4249](https://github.com/bbatsov/rubocop/issues/4249): Handle multiple assignment in `Rails/RelativeDateConstant`. ([@bbatsov][])
* [#4250](https://github.com/bbatsov/rubocop/issues/4250): Improve a bit the Ruby code detection config. ([@bbatsov][])
* [#4283](https://github.com/bbatsov/rubocop/issues/4283): Fix `Style/EmptyCaseCondition` autocorrect bug - when first `when` branch includes comma-delimited alternatives. ([@ilansh][])
* [#4268](https://github.com/bbatsov/rubocop/issues/4268): Handle end-of-line comments when autocorrecting Style/EmptyLinesAroundAccessModifier. ([@vergenzt][])
* [#4275](https://github.com/bbatsov/rubocop/issues/4275): Prevent `Style/MethodCallWithArgsParentheses` from blowing up on `yield`. ([@drenmi][])
* [#3969](https://github.com/bbatsov/rubocop/issues/3969): Handle multiline method call alignment for arguments to methods. ([@jonas054][])
* [#4304](https://github.com/bbatsov/rubocop/pull/4304): Allow enabling whole departments when `DisabledByDefault` is `true`. ([@jonas054][])
* [#4264](https://github.com/bbatsov/rubocop/issues/4264): Prevent `Rails/SaveBang` from blowing up when using the assigned variable in a hash. ([@drenmi][])
* [#4310](https://github.com/bbatsov/rubocop/pull/4310): Treat paths containing invalid byte sequences as non-matches. ([@mclark][])
* [#4339](https://github.com/bbatsov/rubocop/pull/4339): Fix false positive in `Security/Eval` cop for multiline string lietral. ([@pocke][])
* [#4339](https://github.com/bbatsov/rubocop/pull/4339): Fix false negative in `Security/Eval` cop for `Binding#eval`. ([@pocke][])

## 0.48.1 (2017-04-03)

### Changes

* [#4219](https://github.com/bbatsov/rubocop/issues/4219): Add a link to style guide for `Style/IndentationConsistency` cop. ([@pocke][])
* [#4168](https://github.com/bbatsov/rubocop/issues/4168): Removed `-n` option. ([@sadovnik][])
* [#4039](https://github.com/bbatsov/rubocop/pull/4039): Change `Style/PercentLiteralDelimiters` default configuration to match Style Guide update. ([@drenmi][])
* [#4235](https://github.com/bbatsov/rubocop/pull/4235): Improved copy of offense message in `Lint/AmbiguousBlockAssociation` cop. ([@smakagon][])

### Bug fixes

* [#4171](https://github.com/bbatsov/rubocop/pull/4171): Prevent `Rails/Blank` from breaking when RHS of `or` is a naked falsiness check. ([@drenmi][])
* [#4189](https://github.com/bbatsov/rubocop/pull/4189): Make `Lint/AmbiguousBlockAssociation` aware of lambdas passed as arguments. ([@drenmi][])
* [#4179](https://github.com/bbatsov/rubocop/pull/4179): Prevent `Rails/Blank` from breaking when LHS of `or` is a naked falsiness check. ([@rrosenblum][])
* [#4172](https://github.com/bbatsov/rubocop/pull/4172): Fix false positives in `Style/MixinGrouping` cop. ([@drenmi][])
* [#4185](https://github.com/bbatsov/rubocop/pull/4185): Make `Lint/NestedMethodDefinition` aware of `#*_exec` class of methods. ([@drenmi][])
* [#4197](https://github.com/bbatsov/rubocop/pull/4197): Fix false positive in `Style/RedundantSelf` cop with parallel assignment. ([@drenmi][])
* [#4199](https://github.com/bbatsov/rubocop/issues/4199): Fix incorrect auto correction in `Style/SymbolArray` and `Style/WordArray` cop. ([@pocke][])
* [#4218](https://github.com/bbatsov/rubocop/pull/4218): Make `Lint/NestedMethodDefinition` aware of class shovel scope. ([@drenmi][])
* [#4198](https://github.com/bbatsov/rubocop/pull/4198): Make `Lint/AmbguousBlockAssociation` aware of operator methods. ([@drenmi][])
* [#4152](https://github.com/bbatsov/rubocop/pull/4152): Make `Style/MethodCallWithArgsParentheses` not require parens on setter methods. ([@drenmi][])
* [#4226](https://github.com/bbatsov/rubocop/pull/4226): Show in `--help` output that `--stdin` takes a file name argument. ([@jonas054][])

lib/rubbycop/config.rb
# frozen_string_literal: true

require 'pathname'

# rubbycop:disable Metrics/ClassLength

module RubbyCop
  # This class represents the configuration of the RubbyCop application
  # and all its cops. A Config is associated with a YAML configuration
  # file from which it was read. Several different Configs can be used
  # during a run of the rubbycop program, if files in several
  # directories are inspected.
  class Config
    include PathUtil

    COMMON_PARAMS = %w[Exclude Include Severity
                       AutoCorrect StyleGuide Details].freeze
    # 2.1 is the oldest officially supported Ruby version.
    DEFAULT_RUBY_VERSION = 2.1
    KNOWN_RUBIES = [1.9, 2.0, 2.1, 2.2, 2.3, 2.4].freeze
    DEFAULT_RAILS_VERSION = 5.0
    OBSOLETE_COPS = {
      'Style/TrailingComma' =>
        'The `Style/TrailingComma` cop no longer exists. Please use ' \
        '`Style/TrailingCommaInLiteral` and/or ' \
        '`Style/TrailingCommaInArguments` instead.',
      'Rails/DefaultScope' =>
        'The `Rails/DefaultScope` cop no longer exists.',
      'Style/SingleSpaceBeforeFirstArg' =>
        'The `Style/SingleSpaceBeforeFirstArg` cop has been renamed to ' \
        '`Layout/SpaceBeforeFirstArg`.',
      'Lint/SpaceBeforeFirstArg' =>
        'The `Lint/SpaceBeforeFirstArg` cop has been removed, since it was a ' \
        'duplicate of `Layout/SpaceBeforeFirstArg`. Please use ' \
        '`Layout/SpaceBeforeFirstArg` instead.',
      'Layout/SpaceAfterControlKeyword' =>
        'The `Layout/SpaceAfterControlKeyword` cop has been removed. Please ' \
        'use `Layout/SpaceAroundKeyword` instead.',
      'Layout/SpaceBeforeModifierKeyword' =>
        'The `Layout/SpaceBeforeModifierKeyword` cop has been removed. ' \
        'Please use `Layout/SpaceAroundKeyword` instead.',
      'Style/SpaceAfterControlKeyword' =>
        'The `Style/SpaceAfterControlKeyword` cop has been removed. Please ' \
        'use `Layout/SpaceAroundKeyword` instead.',
      'Style/SpaceBeforeModifierKeyword' =>
        'The `Style/SpaceBeforeModifierKeyword` cop has been removed. Please ' \
        'use `Layout/SpaceAroundKeyword` instead.',
      'Style/MethodCallParentheses' =>
        'The `Style/MethodCallParentheses` cop has been renamed to ' \
          '`Style/MethodCallWithoutArgsParentheses`.',
      'Lint/Eval' =>
        'The `Lint/Eval` cop has been renamed to `Security/Eval`.',
      'Style/DeprecatedHashMethods' =>
        'The `Style/DeprecatedHashMethods` cop has been renamed to ' \
          '`Style/PreferredHashMethods`.'
    }.freeze

    OBSOLETE_PARAMETERS = [
      {
        cop: 'Layout/SpaceAroundOperators',
        parameter: 'MultiSpaceAllowedForOperators',
        alternative: 'If your intention was to allow extra spaces ' \
                     'for alignment, please use AllowForAlignment: ' \
                     'true instead.'
      },
      {
        cop: 'Style/SpaceAroundOperators',
        parameter: 'MultiSpaceAllowedForOperators',
        alternative: 'If your intention was to allow extra spaces ' \
                     'for alignment, please use AllowForAlignment: ' \
                     'true instead.'
      },
      {
        cop: 'AllCops',
        parameter: 'RunRailsCops',
        alternative: "Use the following configuration instead:\n" \
                     "Rails:\n  Enabled: true"
      },
      {
        cop: 'Layout/CaseIndentation',
        parameter: 'IndentWhenRelativeTo',
        alternative: '`IndentWhenRelativeTo` has been renamed to ' \
                     '`EnforcedStyle`'
      },
      {
        cop: 'Lint/BlockAlignment',
        parameter: 'AlignWith',
        alternative: '`AlignWith` has been renamed to ' \
                     '`EnforcedStyleAlignWith`'
      },
      {
        cop: 'Lint/EndAlignment',
        parameter: 'AlignWith',
        alternative: '`AlignWith` has been renamed to ' \
                     '`EnforcedStyleAlignWith`'
      },
      {
        cop: 'Lint/DefEndAlignment',
        parameter: 'AlignWith',
        alternative: '`AlignWith` has been renamed to ' \
                     '`EnforcedStyleAlignWith`'
      },
      {
        cop: 'Rails/UniqBeforePluck',
        parameter: 'EnforcedMode',
        alternative: '`EnforcedMode` has been renamed to ' \
                     '`EnforcedStyle`'
      }
    ].freeze

    attr_reader :loaded_path

    def initialize(hash = {}, loaded_path = nil)
      @loaded_path = loaded_path
      @for_cop = Hash.new do |h, cop|
        qualified_cop_name = Cop::Cop.qualified_cop_name(cop, loaded_path)
        cop_options = self[qualified_cop_name] || {}
        cop_options['Enabled'] = enable_cop?(qualified_cop_name, cop_options)
        h[cop] = cop_options
      end
      @hash = hash
    end

    def [](key)
      @hash[key]
    end

    def []=(key, value)
      @hash[key] = value
    end

    def delete(key)
      @hash.delete(key)
    end

    def each(&block)
      @hash.each(&block)
    end

    def key?(key)
      @hash.key?(key)
    end

    def keys
      @hash.keys
    end

    def map(&block)
      @hash.map(&block)
    end

    def merge(other_hash)
      @hash.merge(other_hash)
    end

    def to_h
      @hash
    end

    def to_hash
      @hash
    end

    def to_s
      @to_s ||= @hash.to_s
    end

    def make_excludes_absolute
      each do |key, _|
        validate_section_presence(key)
        next unless self[key]['Exclude']

        self[key]['Exclude'].map! do |exclude_elem|
          if exclude_elem.is_a?(String) && !absolute?(exclude_elem)
            File.expand_path(File.join(base_dir_for_path_parameters,
                                       exclude_elem))
          else
            exclude_elem
          end
        end
      end
    end

    def add_excludes_from_higher_level(highest_config)
      return unless highest_config.for_all_cops['Exclude']

      excludes = for_all_cops['Exclude'] ||= []
      highest_config.for_all_cops['Exclude'].each do |path|
        unless path.is_a?(Regexp) || absolute?(path)
          path = File.join(File.dirname(highest_config.loaded_path), path)
        end
        excludes << path unless excludes.include?(path)
      end
    end

    def deprecation_check
      %w[Exclude Include].each do |key|
        plural = "#{key}s"
        next unless for_all_cops[plural]

        for_all_cops[key] = for_all_cops[plural] # Stay backwards compatible.
        for_all_cops.delete(plural)
        yield "AllCops/#{plural} was renamed to AllCops/#{key}"
      end
    end

    def for_cop(cop)
      @for_cop[cop.respond_to?(:cop_name) ? cop.cop_name : cop]
    end

    def for_all_cops
      @for_all_cops ||= self['AllCops'] || {}
    end

    def validate
      # Don't validate RubbyCop's own files. Avoids infinite recursion.
      base_config_path = File.expand_path(File.join(ConfigLoader::RUBOCOP_HOME,
                                                    'config'))
      return if File.expand_path(loaded_path).start_with?(base_config_path)

      valid_cop_names, invalid_cop_names = keys.partition do |key|
        ConfigLoader.default_configuration.key?(key)
      end

      reject_obsolete_cops_and_parameters
      warn_about_unrecognized_cops(invalid_cop_names)
      check_target_ruby
      validate_parameter_names(valid_cop_names)
      validate_enforced_styles(valid_cop_names)
      reject_mutually_exclusive_defaults
    end

    def file_to_include?(file)
      relative_file_path = path_relative_to_config(file)

      # Optimization to quickly decide if the given file is hidden (on the top
      # level) and can not be matched by any pattern.
      is_hidden = relative_file_path.start_with?('.') &&
                  !relative_file_path.start_with?('..')
      return false if is_hidden && !possibly_include_hidden?

      absolute_file_path = File.expand_path(file)

      patterns_to_include.any? do |pattern|
        match_path?(pattern, relative_file_path) ||
          match_path?(pattern, absolute_file_path)
      end
    end

    # Returns true if there's a chance that an Include pattern matches hidden
    # files, false if that's definitely not possible.
    def possibly_include_hidden?
      return @possibly_include_hidden if defined?(@possibly_include_hidden)

      @possibly_include_hidden = patterns_to_include.any? do |s|
        s.is_a?(Regexp) || s.start_with?('.') || s.include?('/.')
      end
    end

    def file_to_exclude?(file)
      file = File.expand_path(file)
      patterns_to_exclude.any? do |pattern|
        match_path?(pattern, file)
      end
    end

    def patterns_to_include
      for_all_cops['Include']
    end

    def patterns_to_exclude
      for_all_cops['Exclude']
    end

    def path_relative_to_config(path)
      relative_path(path, base_dir_for_path_parameters)
    end

    # Paths specified in configuration files starting with .rubbycop are
    # relative to the directory where that file is. Paths in other config files
    # are relative to the current directory. This is so that paths in
    # config/default.yml, for example, are not relative to RubbyCop's config
    # directory since that wouldn't work.
    def base_dir_for_path_parameters
      @base_dir_for_path_parameters ||=
        if File.basename(loaded_path).start_with?('.rubbycop') &&
           loaded_path != File.join(Dir.home, ConfigLoader::DOTFILE)
          File.expand_path(File.dirname(loaded_path))
        else
          Dir.pwd
        end
    end

    def target_ruby_version
      @target_ruby_version ||=
        if for_all_cops['TargetRubyVersion']
          @target_ruby_version_source = :rubbycop_yml

          for_all_cops['TargetRubyVersion']
        elsif File.file?('.ruby-version') &&
              /\A(ruby-)?(?<version>\d+\.\d+)/ =~ File.read('.ruby-version')

          @target_ruby_version_source = :dot_ruby_version

          version.to_f
        else
          DEFAULT_RUBY_VERSION
        end
    end

    def target_rails_version
      @target_rails_version ||=
        for_all_cops.fetch('TargetRailsVersion', DEFAULT_RAILS_VERSION)
    end

    private

    def warn_about_unrecognized_cops(invalid_cop_names)
      invalid_cop_names.each do |name|
        if name == 'Syntax'
          raise ValidationError,
                "configuration for Syntax cop found in #{loaded_path}\n" \
                'This cop cannot be configured.'
        end

        # There could be a custom cop with this name. If so, don't warn
        next if Cop::Cop.registry.contains_cop_matching?([name])

        warn Rainbow("Warning: unrecognized cop #{name} found in " \
                     "#{loaded_path}").yellow
      end
    end

    def validate_section_presence(name)
      return unless key?(name) && self[name].nil?
      raise ValidationError, "empty section #{name} found in #{loaded_path}"
    end

    def validate_parameter_names(valid_cop_names)
      valid_cop_names.each do |name|
        validate_section_presence(name)
        self[name].each_key do |param|
          next if COMMON_PARAMS.include?(param) ||
                  ConfigLoader.default_configuration[name].key?(param)

          warn Rainbow("Warning: unrecognized parameter #{name}:#{param} " \
                       "found in #{loaded_path}").yellow
        end
      end
    end

    def validate_enforced_styles(valid_cop_names)
      valid_cop_names.each do |name|
        styles = self[name].select { |key, _| key.start_with?('Enforced') }

        styles.each do |style_name, style|
          supported_key = RubbyCop::Cop::Util.to_supported_styles(style_name)
          valid = ConfigLoader.default_configuration[name][supported_key]
          next unless valid
          next if valid.include?(style)

          msg = "invalid #{style_name} '#{style}' for #{name} found in " \
            "#{loaded_path}\n" \
            "Valid choices are: #{valid.join(', ')}"
          raise ValidationError, msg
        end
      end
    end

    def reject_obsolete_cops_and_parameters
      messages = [
        obsolete_cops,
        obsolete_parameters
      ].flatten.compact
      return if messages.empty?

      raise ValidationError, messages.join("\n")
    end

    def obsolete_parameters
      OBSOLETE_PARAMETERS.map do |params|
        obsolete_parameter_message(params[:cop], params[:parameter],
                                   params[:alternative])
      end
    end

    def obsolete_parameter_message(cop, parameter, alternative)
      return unless self[cop] && self[cop].key?(parameter)

      "obsolete parameter #{parameter} (for #{cop}) " \
        "found in #{loaded_path}" \
        "\n#{alternative}"
    end

    def obsolete_cops
      OBSOLETE_COPS.map do |cop_name, message|
        next unless key?(cop_name) || key?(Cop::Badge.parse(cop_name).cop_name)
        message + "\n(obsolete configuration found in #{loaded_path}, please" \
                   ' update it)'
      end
    end

    def check_target_ruby
      return if KNOWN_RUBIES.include?(target_ruby_version)

      msg = "Unknown Ruby version #{target_ruby_version.inspect} found "

      msg +=
        case @target_ruby_version_source
        when :dot_ruby_version
          'in `.ruby-version`.'
        when :rubbycop_yml
          "in `TargetRubyVersion` parameter (in #{loaded_path})." \
        end

      msg += "\nKnown versions: #{KNOWN_RUBIES.join(', ')}"

      raise ValidationError, msg
    end

    def reject_mutually_exclusive_defaults
      disabled_by_default = for_all_cops['DisabledByDefault']
      enabled_by_default = for_all_cops['EnabledByDefault']
      return unless disabled_by_default && enabled_by_default

      msg = 'Cops cannot be both enabled by default and disabled by default'
      raise ValidationError, msg
    end

    def enable_cop?(qualified_cop_name, cop_options)
      cop_department, cop_name = qualified_cop_name.split('/')
      department = cop_name.nil?

      unless department
        department_options = self[cop_department]
        if department_options && department_options.fetch('Enabled') == false
          return false
        end
      end

      cop_options.fetch('Enabled', true)
    end
  end
end
